README.md
# et_upload-cookbook

A cookbook to provision an SFTP server which a collection of chroot jailed users. Primarily aimed at situations where end users need simple, but limited, SCP/SFTP access to provide data (e.g., automated importing).

## Supported Platforms

* Ubuntu 12.04

## Attributes

Key                      | Type    | Description              | Default
---                      | ----    | -----------              | -------
`['openssh']['server']['port']`                    | Array  | Ports OpenSSH listens on           | `%w(22 43827)`
`['openssh']['server']['permit_root_login']`       | String | Allow remote root logins           | `'no'`
`['openssh']['server']['password_authentication']` | String | Allow password logins              | `'yes'`
`['openssh']['server']['subsystem']`               | String | Set a subsystem for OpenSSH        | `'sftp /usr/lib/sftp-server'`
`['openssh']['server']['match']`                   | Hash   | Provide a match config for OpenSSH | see below

```ruby
set['openssh']['server']['match'] = {
  'Group uploadonly' => {
    'chroot_directory' => '%h',
    'force_command' => 'internal-sftp',
    'allow_tcp_forwarding' => 'no'
  }
}
```

## Files

### recipe/scripts

Performs several setup steps that are required for processing uploads and exports.  This script does the following:

1. Setup the rotation schedule for /var/log/process\_uploads.log and /var/log/process\_scheduled\_exports.log
2. Create /opt/evertrue/config.xml
3. Setup cron schedules for processing uploads and exports and removing old upload and export files

### recipe/users

Manages folder structure of all upload users.  This script does the following:

1. Load the upload.json content from the users databag
2. For each user do the following:
3. Determine the home folder of the user.  For most users the folder will be /mnt/dev0/evertrue/users/{username}.  However, if the parent\_user and partner values have been set for the user then the home folder will be a subfolder under the parent\_user named for the partner.
4. If the user's action is remove then delete the user's home directory and all files, otherwise do the following:
  5. Create the OS user and set the home folder
  6. Create the user's home folder
  7. If the user is using SSH keys then store those in the .ssh folder
  8. Create the uploads and exports folders under the home folder

An example user object from the upload.json file:

```
  "givingtreedemo": {
    "uid": 10150,
    "ssh_keys": [],
    "comment": "Chelsea Leavitt GivingTree Demo",
    "password": "abc"
  }
```

An example parter user object:

```
  "givingtreedemo-hustle": {
    "uid": 10663,
    "ssh_keys": [],
    "parent_user": "givingtreedemo",
    "partner": "hustle",
    "comment": "GivingTree Demo Hustle Access",
    "password": "abc"
  }
```


### files/default/process\_uploads

Sends files from the uploads directory for import processing.  This script does the following:

1. Load /opt/evertrue/config.xml
2. For each value in unames, do the following:
3. Get all file names stored under any uploads folder under the uname home folder
4. Ignore any upload file that does not have a csv, gz or zip extension.
5. Determine the type of import from the name of the upload file
6. Copy the upload file to the org's S3 folder
7. Check that a mapping exists for the upload file's headers
8. Post the upload file to the importer
9. If a mapping exists then queue the import file for processing if `ET.Importer.IngestionMode = AutoIngest`
10. If a mapping does not exist then notify support and set `ET.Importer.IngestionMode = NotifyOnly`

Logs for this script are saved to /var/log/process\_uploads.log

### files/default/process\_scheduled\_exports

Copies scheduled export files to the org's SFTP folder for download.  This script does the following:

1. Load /opt/evertrue/config.xml
2. For each value in unames, do the following:
3. Get the oid from the uname value.  Note that not all unames map directly to an org slug
4. Get the latest scheduled exports for the oid via /contacts/v2/exports/latest-scheduled
5. Download each export file via /contacts/v2/exports
6. Save the export file to the uname home exports folder

Logs for this script are saved to /var/log/process\_scheduled\_exports.log

## Usage

### et_upload::default

Include `et_upload` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[et_upload::default]"
  ]
}
```

For testing purposes, the users upload data bag item exists. The password for each user is `password`, salted & encrypted to best resemble a real password & allow for logging in via SFTP to do manual testing of SFTP functionality.

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: EverTrue, Inc. (<jeff@evertrue.com>)


CHANGELOG.md
et_upload cookbook CHANGELOG
============================
This file is used to list changes made in each version of the et_upload cookbook.

5.1.28 (2021-03-04)
-------------------
* Updated process_uploads to ignore reports folder

5.1.27 (2021-03-01)
-------------------
* Added a reporting folder to all clients

5.1.22 (2020-01-15)
------------------
* Need to shellword escape the file name in gzip command

5.1.22 (2020-01-13)
------------------
* Use gzip command for compressing files
* Ensure files are in the uploads folders and not a subfolder

v5.1.21 (2020-01-13)
------------------
* Added partner upload support

v5.1.19 (2019-03-07)
------------------
* Added new solicitors file type to upload process

v5.1.17 (2019-02-07)
------------------
* Upload files in the order we receive them

v5.1.15 (2019-01-11)
------------------
* Run process_uploads every 2 minutes now that missing org files have been removed

v5.1.14 (2019-01-11)
------------------
* Fix for capturing mtime before deleting file

v5.1.13 (2019-01-11)
------------------
* Remove files from orgs that are not found

v5.1.12 (2019-01-11)
------------------
* Run process_uploads every 5 minutes

v5.1.11 (2018-12-21)
------------------

* Notify customer support about invalid zip files

v5.1.10 (2018-06-19)
------------------

* Use newer version of gem

v5.1.9 (2018-06-19)
------------------

* Update evertrue group permissions on uploads and exports

v5.1.8 (2018-05-31)
------------------

* Revert group change on the home dir

v5.1.7 (2018-05-30)
------------------

* Fix .ssh permission

v5.1.6 (2018-05-30)
------------------

* Give evertrue user permission to see inside upload users' home dir

v5.1.5 (2018-04-09)
------------------

* Blake blew out a return statement for the s3_filename, put that back
  in

v5.1.3 (2018-03-19)
------------------

* File naming convention can be singular

v5.1.2 (2018-03-03)
------------------

* Only download export once

v5.1.1 (2017-08-11)
------------------

* Add special exception for yale

v5.1.0 (2017-08-11)
------------------

* Process assignment imports

v5.0.5 (2017-07-11)
------------------

* Log scheduled export to a different file

v5.0.4 (2017-07-11)
------------------

* Run scheduled export job every 10 minutes
* Clean scheduled export every 90 days

v5.0.3 (2017-07-11)
------------------

* Remove extension from process scheduled export.

v5.0.2 (2017-07-11)
------------------

* Add process scheduled export to the cookbook file.

v5.0.1 (2017-07-11)
------------------

* Process Scheduled Exports: Add script to process latest scheduled exports from CAPI per org.

v4.0.0 (2017-01-25)
------------------

* Process uploads: Make sure the algorithm for REMOVING slug random numbers matches the one for adding them

v3.0.7 (2016-12-13)
------------------

* Process uploads: Make sure the algorithm for REMOVING slug random numbers matches the one for adding them

v3.0.6 (2016-09-20)
------------------

* Add reboot coordinator exemption

v3.0.5 (2016-08-09)
------------------

* Have upload scripts use their own sentry dsn

v3.0.4 (2016-7-01)
------------------

* Fix bug with pagerduty key

v3.0.3 (2016-6-28)
------------------

* Add support for interaction imports

v3.0.2 (2016-4-12)
------------------

* Add support for full gift imports

v3.0.1 (2016-4-12)
------------------

* Add email notification support for genius@evertrue.com

v3.0.0 (2016-03-28)
-------------------

* Upgrade to Ruby 2.0 for upload processing scripts

v2.3.7 (2016-03-25)
-------------------

* Use sudo to allow group evertrue users to run chef-client and restart the service

v2.3.6 (2016-02-26)
-------------------

* Don't fail the process_uploads script for API response codes of 400,
  since it keeps other files from being processed

v2.3.5 (2015-09-25)
-------------------

* Re-enable pagerduty for the process uploads script
* Send support emails to onboarding instead of Jenna
* Use env var `DEBUG_EMAIL` instead of `alex@evertrue.com`

v2.3.4 (2015-09-10)
-------------------

* Fixes authentication issue when hitting new DNA endpoint

v2.3.3 (2015-09-10)
-------------------

* Use non-legacy DNA endpoint for org setting values

v2.3.2 (2015-09-08)
-------------------

* Rotate logs weekly to get a better historical view of import statuses

v2.3.1 (2015-08-29)
-------------------

* Fix bug causing script to fail in production

v2.3.0 (2015-08-28)
-------------------

* Many improvements to the process uploads script (mostly more logging and alerting)

v2.2.2 (2015-08-21)
-------------------

* Resolve script error that prevents DNA updates for unrecognized csv
  headers

v2.2.1 (2015-08-05)
-------------------

* `s/work_dir/archive_dir/`

v2.2.0 (2015-08-05)
-------------------

* Refactor the process uploads script
* Move working directory and uploads (users) directories to ephemeral storage
* Add support for header changes

v2.1.2 (2015-07-24)
-------------------

* Add tests for general Contact import path in process_uploads

v2.1.1 (2015-07-23)
-------------------

* Filter out users with action property set to 'remove'

v2.1.0 (2015-07-23)
-------------------

* Allow Transactional Gift file processing through the process_uploads
  script

v2.0.18 (2015-07-23)
-------------------

* Refactor process_upload script and surrounding infrastructure
* Testing on Ubuntu 14.4
* Add integration test for process_upload script
* Many minor bug fixes and improvements

v2.0.17 (2015-04-27)
-------------------

* Handle user deletes

v2.0.16 (2015-04-27)
-------------------

* Remove legacy code path

v2.0.15 (2015-04-13)
-------------------

* Use greater-than version constraint for cron cookbook

v2.0.14 (2015-02-24)
-------------------

* Fix hostname for hb

v2.0.13 (2014-08-13)
-------------------

* Add notification when queueing a job

v2.0.12 (2014-07-30)
-------------------

* Escape spaces only for shell exec

v2.0.11 (2014-06-26)
-------------------

* Escape spaces in filenames


attributes/default.rb
# ssh config
# set['openssh']['client']['host'] = '*'
# set['openssh']['client']['send_env'] = 'LANG LC_*'
# set['openssh']['client']['hash_known_hosts'] = 'yes'
# set['openssh']['client']['gssapi_authentication'] = 'yes'
# set['openssh']['client']['gssapi_delegate_credentials'] = 'no'

# sshd config
set['openssh']['server']['port'] = %w(22 43827)
# set['openssh']['server']['protocol'] = '2'
# set['openssh']['server']['host_key_rsa'] = '/etc/ssh/ssh_host_rsa_key'
# set['openssh']['server']['host_key_dsa'] = '/etc/ssh/ssh_host_dsa_key'
# set['openssh']['server']['host_key_ecdsa'] = '/etc/ssh/ssh_host_ecdsa_key'
# set['openssh']['server']['use_privilege_separation'] = 'yes'
# set['openssh']['server']['key_regeneration_interval'] = '3600'
# set['openssh']['server']['server_key_bits'] = '768'
# set['openssh']['server']['syslog_facility'] = 'AUTH'
# set['openssh']['server']['log_level'] = 'INFO'
# set['openssh']['server']['login_grace_time'] = '120'
set['openssh']['server']['permit_root_login'] = 'no'
# set['openssh']['server']['strict_modes'] = 'yes'
# set['openssh']['server']['r_s_a_authentication'] = 'yes'
# set['openssh']['server']['pubkey_authentication'] = 'yes'
# set['openssh']['server']['ignore_rhosts'] = 'yes'
# set['openssh']['server']['rhosts_r_s_a_authentication'] = 'no'
# set['openssh']['server']['hostbased_authentication'] = 'no'
# set['openssh']['server']['permit_empty_passwords'] = 'no'
# set['openssh']['server']['challenge_response_authentication'] = 'no'
set['openssh']['server']['password_authentication'] = 'yes'
# set['openssh']['server']['x11_forwarding'] = 'yes'
# set['openssh']['server']['x11_display_offset'] = '10'
# set['openssh']['server']['print_motd'] = 'no'
# set['openssh']['server']['print_last_log'] = 'yes'
# set['openssh']['server']['t_c_p_keep_alive'] = 'yes'
# set['openssh']['server']['accept_env'] = 'LANG LC_*'
set['openssh']['server']['subsystem'] = 'sftp /usr/lib/sftp-server'
# set['openssh']['server']['use_p_a_m'] = 'yes'
# set['openssh']['server']['use_d_n_s'] = 'no'

set['openssh']['server']['match'] = {
  'Group uploadonly' => {
    'chroot_directory' => '%h',
    'force_command' => 'internal-sftp',
    'allow_tcp_forwarding' => 'no'
  }
}

override['reboot_coordinator']['reboot_permitted'] = false
